test/conftest.py

```diff
@@ -202,33 +257,57 @@ def image_url():
     return f'file://{path}'
 
 
-def setup_db(db, **kwargs):
+def create_db(CFG, **kwargs):
     # TODO: support more parameters to control the setup
+    db = build_datalayer(CFG)
+    if kwargs.get('empty', False):
+        return db
+
     add_encoders(db)
-    n_data = kwargs.get('n_data', GLOBAL_TEST_N_DATA_POINTS)
-    add_random_data(db, number_data_points=n_data)
+
+    # prepare data
+    n_data = kwargs.get('n_data', LOCAL_TEST_N_DATA_POINTS)
+    is_mongodb_bachend = isinstance(db.databackend, MongoDataBackend)
+    if is_mongodb_bachend:
+        add_random_data_to_mongo_db(db, number_data_points=n_data)
+    else:
+        add_random_data_to_sql_db(db, number_data_points=n_data)
+
+    # prepare models
     if kwargs.get('add_models', True):
         add_models(db)
+
+    # prepare vector index
     if kwargs.get('add_vector_index', True):
         add_vector_index(db)
 
-
-@pytest.fixture(scope='session')
-def db() -> Datalayer:
-    db = build_datalayer(CFG, data_backend=MONGOMOCK_URI)
-    setup_db(db)
     return db
 
 
 @pytest.fixture
-def local_db(request) -> Datalayer:
-    db = build_datalayer(CFG, data_backend=MONGOMOCK_URI)
-    setup_config = getattr(request, 'param', {'n_data': LOCAL_TEST_N_DATA_POINTS})
-    setup_db(db, **setup_config)
-    return db
-
+def db(request, monkeypatch) -> Iterator[Datalayer]:
+    # TODO: Use pre-defined config instead of dict here
+    param = request.param if hasattr(request, 'param') else DBConfig.mongodb
+    if isinstance(param, dict):
+        # e.g. @pytest.mark.parametrize("db", [DBConfig.sqldb], indirect=True)
+        setup_config = param.copy()
+    elif isinstance(param, tuple):
+        # e.g. @pytest.mark.parametrize("db", [(DBConfig.sqldb, {'n_data': 10})], indirect=True) # noqa: E501
+        setup_config = param[0].copy()
+        setup_config.update(param[1] or {})
+    else:
+        raise ValueError(f'Unsupported param type: {type(param)}')
+
+    monkeypatch.setattr(CFG, 'data_backend', setup_config['data_backend'])
+
+    db = create_db(CFG, **setup_config)
+    yield db
```

Comment:
Configure the db fixture in test cases using the predefined DBConfig.

Standardize fixture creation using configuration settings, with support for MongoDB and SQL databases.

Then, every test case can use the same fixture, yet they can still have different setup behaviors.

- DBConfig.mongodb_empty
- DBConfig.sqldb_empty
- DBConfig.mongodb (include: encoder, data, model, vector_index)
- DBConfig.sqldb (include: encoder, data, model, vector_index)

The above four are the most commonly used configurations. If there are more frequently used database preset configurations, new configurations can also be preset, as follows:

- DBConfig.mongodb_no_vector_index (include: encoder, data, model)
- DBConfig.sqldb_no_vector_index (include: encoder, data, model)
- DBConfig.mongodb_data (include: encoder, data)
- DBConfig.sqldb_data (include: encoder, data)


For example:

**default setup behavior with using `mongodb`**

```python
def test_delete_many(db):
    ...
```


**empty mongodb**

```python
 @pytest.mark.parametrize("db", [DBConfig.mongodb_empty], indirect=True)
 def test_execute_insert_and_find(db):
    ...
```

**empty `mongodb` and empty `sqldb`**

```python
@pytest.mark.parametrize(
     "db", [DBConfig.mongodb_empty, DBConfig.sqldb_empty], indirect=True
 )
 def test_add_version(db):
    ....
```

**do not add vector_index and add 500 data points**

```python
pytest.mark.parametrize(
     "db",
     [
         (DBConfig.mongodb_no_vector_index, {'n_data': 500}),
         (DBConfig.sqldb_no_vector_index, {'n_data': 500}),
     ],
     indirect=True,
 )
 def test_dataset(db):
    ...
```

or use config param to do this

```python
pytest.mark.parametrize(
     "db",
     [
         (DBConfig.mongodb, {'add_vector_index': False, 'n_data': 500}),
         (DBConfig.sqldb, {'add_vector_index': False, 'n_data': 500}),
     ],
     indirect=True,
 )
 def test_dataset(db):
    ...
```

**do not add vector_index and add default size (5) data points**

```python
pytest.mark.parametrize(
     "db",
     [
         DBConfig.mongodb_no_vector_index,
         DBConfig.sqldb_no_vector_index,
     ],
     indirect=True,
 )
 def test_dataset(db):
    ...
```

Reply:
https://github.com/jieguangzhou/superduperdb/blob/unittests/sqllite/test/db_config.py